src/detect-flowbits.c

```diff
@@ -809,6 +901,470 @@ static void DetectFlowbitsAnalyzeDump(const DetectEngineCtx *de_ctx,
+                                       ? sigmatch_table[s->init_data->prefilter_sm->type].name
+                                       : "none";
+            SCLogDebug("flowbit: %u => rejected sid %u (iid:%u). No prefilter or prefilter not "
+                       "flowbits (%p, %s, %d)",
+                    fb->idx, s->id, sig_iid, s->init_data->prefilter_sm, name,
+                    s->init_data->prefilter_sm ? s->init_data->prefilter_sm->type : -1);
+#endif
+            continue;
+        }
+
+        /* only add sids that match our bit */
+        const DetectFlowbitsData *fs_fb =
+                (const DetectFlowbitsData *)s->init_data->prefilter_sm->ctx;
+        if (fs_fb->idx != fb->idx) {
+            SCLogDebug(
+                    "flowbit: %u => rejected sid %u (iid:%u). Sig prefilters on different bit %u",
+                    fb->idx, s->id, sig_iid, fs_fb->idx);
+            continue;
+        }
+
+        bool dup = false;
+        for (uint32_t x = 0; x < add->rule_id_cnt; x++) {
+            if (add->rule_id[x] == sig_iid) {
+                dup = true;
+            }
+        }
+
+        if (!dup) {
+            if (add->rule_id_cnt < add->rule_id_size) {
+                add->rule_id[add->rule_id_cnt++] = sig_iid;
+            } else {
+                uint32_t *ptr = SCRealloc(
+                        add->rule_id, (add->rule_id_size + BLOCK_SIZE) * sizeof(uint32_t));
+                if (ptr == NULL) {
+                    return -1;
+                }
+                add->rule_id = ptr;
+                add->rule_id_size += BLOCK_SIZE;
+                add->rule_id[add->rule_id_cnt++] = sig_iid;
+            }
+            added = 1;
+            SCLogDebug("flowbit: %u => accepted sid %u (iid:%u)", fb->idx, s->id, sig_iid);
+        }
+    }
+    return added;
+}
+
+/* TODO shouldn't add sids for which Signature::num is < our num. Is this possible after sorting? */
+
+/** \brief For set/toggle flowbits, build "set" post-rule-match engine
+ *
+ *  For set/toggle flowbits, a special post-rule-match engine is constructed
+ *  to update the running match array during rule matching.
+ */
+static int AddBitSetToggle(const DetectEngineCtx *de_ctx, struct FBAnalyzer *fba,
+        struct PrefilterEngineFlowbits *ctx, const DetectFlowbitsData *fb, const Signature *s)
+{
+    PrefilterFlowbit x;
+    memset(&x, 0, sizeof(x));
+    x.id = fb->idx;
+    PrefilterFlowbit *pfb = PFB_RB_FIND(&ctx->fb_tree, &x);
+    if (pfb == NULL) {
+        PrefilterFlowbit *add = SCCalloc(1, sizeof(*add));
+        if (add == NULL)
+            return -1;
+
+        add->id = fb->idx;
+        add->rule_id_size = NextMultiple(fba->array[fb->idx].isset_sids_idx, BLOCK_SIZE);
+        add->rule_id = SCCalloc(1, add->rule_id_size * sizeof(uint32_t));
+        if (add->rule_id == NULL) {
+            SCFree(add);
+            return -1;
+        }
+
+        if (!AddIssetSidsForBit(de_ctx, fba, fb, add)) {
```

Comment:
This could return -1 on error as well, which is not caught here.

Reply:
good catch